subcommand/tls-init/command.go

```diff
@@ -0,0 +1,368 @@
+package tls_init
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/consul-k8s/helper/cert"
+	"github.com/hashicorp/consul-k8s/subcommand"
+	"github.com/hashicorp/consul-k8s/subcommand/common"
+	"github.com/hashicorp/consul-k8s/subcommand/flags"
+	"github.com/hashicorp/go-hclog"
+	"github.com/mitchellh/cli"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+var (
+	ca string
+	pk string
+)
+
+type Command struct {
+	UI        cli.Ui
+	clientset kubernetes.Interface
+
+	flags *flag.FlagSet
+	k8s   *flags.K8SFlags
+
+	// flags that support the CA/key as files on disk.
+	caFile  string
+	keyFile string
+
+	// flags that support the CA/key as secrets in Kubernetes.
+	caCertSecret *corev1.Secret
+	caKeySecret  *corev1.Secret
+
+	// flags that dictate the specifications of the created certs.
+	days        int
+	domain      string
+	dc          string
+	dnsnames    flags.AppendSliceValue
+	ipaddresses flags.AppendSliceValue
+
+	// flags that dictate specifics for the secret name and namespace
+	// that are created by the command.
+	k8sNamespace string
+	namePrefix   string
+
+	// log
+	log          hclog.Logger
+	flagLogLevel string
+
+	ctx context.Context
+
+	once sync.Once
+	help string
+}
+
+func (c *Command) Run(args []string) int {
+	c.once.Do(c.init)
+	if err := c.flags.Parse(args); err != nil {
+		if err == flag.ErrHelp {
+			return 0
+		}
+		c.UI.Error(fmt.Sprintf("Failed to parse args: %v", err))
+		return 1
+	}
+
+	if len(c.flags.Args()) > 0 {
+		c.UI.Error("Should have no non-flag arguments.")
+		return 1
+	}
+
+	if err := c.validateFlags(); err != nil {
+		c.UI.Error(err.Error())
+		return 1
+	}
+
+	var err error
+	c.log, err = common.Logger(c.flagLogLevel)
+	if err != nil {
+		c.UI.Error(err.Error())
+		return 1
+	}
+
+	if c.clientset == nil {
+		if err := c.configureKubeClient(); err != nil {
+			c.log.Error("error configuring kubernetes client", "err", err.Error())
```

Comment:
Yeah. I wasn't entirely certain why we had two loggers either. I assumed the UI logger was purely for bootstrapping errors and once the command initialized it's own logger, it would use that everywhere.
I am not the biggest fan of the existing pattern but I don't think i have a strong enough opinion about it one way or the other.
> maybe we should switch to logger for all of these?

You mean the UI logger or the logger that we initialize within the command? What would you recommend doing?